src/simulations/modeling/adapter.py

```diff
@@ -277,20 +277,92 @@ def compare_feature(gene):
             reaction.build_reaction_from_string(equation)
             new_metabolites = set(model.metabolites) - metabolites_before
 
+            # Ensure all metabolites have a compartment. (Check all of the reaction's
+            # metabolites, but presumably only new metabolites will not have a
+            # compartment.)
+            for metabolite in reaction.metabolites.keys():
+                if metabolite.compartment:
+                    continue
+                # Assume BiGG identifier convention and try to parse the compartment
+                # id.
+                if "_" not in metabolite.id:
+                    error = (
+                        f"Metabolite {metabolite.id} is unknown, and we cannot parse a "
+                        "compartment for it."
+                    )
+                    errors.append(error)
+                    logger.error(error)
+                    continue
+                metabolite_id, compartment_id = metabolite.id.rsplit("_", 1)
+                if compartment_id not in model.compartments:
+                    error = (
+                        f"Compartment {compartment_id} does not exist in the model,"
+                        f"but that's what we understand metabolite {metabolite.id} "
+                        f"to exist in."
+                    )
+                    errors.append(error)
+                    logger.error(error)
+                    continue
+                logger.debug(
+                    f"Setting compartment for metabolite {metabolite} to: "
+                    f"{compartment_id}"
+                )
+                metabolite.compartment = compartment_id
+
             operations.append({
                 'operation': 'add',
                 'type': 'reaction',
                 'data': reaction_to_dict(reaction),
             })
 
-            # For all new metabolites, create a demand reactions so that they
-            # may leave the system.
+            # We have to ensure that all new metabolites can leave the system. Create an
+            # extracellular metabolite + exchange reaction (as opposed to just creating
+            # a demand reaction for the intracellular metabolite) because if there are
+            # metabolomics for this metabolite in a later step, our adapter logic always
+            # assumes it exists in the 'e' compartment and that there exists an exchange
+            # reaction.
             for metabolite in new_metabolites:
-                demand_reaction = model.add_boundary(metabolite, type='demand')
+                # Create an extracellular version of the same metabolite
+                if metabolite.compartment == "e":
+                    error = (
```

Comment:
Instead of considering this an error, I would maybe consider things to work as expected already?

Reply:
Indeed, sounds like it's much better to try instead of just failing.